pkg/storage/tsdb/tombstones.go

```diff
@@ -0,0 +1,342 @@
+		State:            state,
+	}
+}
+
+// Uploads a tombstone file to object sotre
+func WriteTombstoneFile(ctx context.Context, bkt objstore.Bucket, userID string, cfgProvider bucket.TenantConfigProvider, tombstone *Tombstone) error {
+	userLogger := util_log.WithUserID(userID, util_log.Logger)
+	userBkt := bucket.NewUserBucketClient(userID, bkt, cfgProvider)
+
+	data, err := json.Marshal(tombstone)
+	if err != nil {
+		return errors.Wrap(err, "serialize tombstone")
+	}
+
+	// Add the state of the request using the filename extension
+	filename := tombstone.RequestID + "." + string(tombstone.State) + ".json"
+
+	fullTombstonePath := path.Join(TombstonePath, filename)
+
+	// Check if the tombstone already exists. Could be the case the same request was made
+	// and is already in the middle of deleting series. Creating a new tombstone would restart
+	// the progress
+	tombstoneExists, err := TombstoneExists(ctx, userBkt, userID, tombstone.RequestID, tombstone.State)
+	if err != nil {
+		level.Error(userLogger).Log("msg", "unable to check if the same tombstone already exists", "requestID", tombstone.RequestID, "err", err)
+	} else if tombstoneExists {
+		return ErrTombstoneAlreadyExists
+	}
+
+	return errors.Wrap(userBkt.Upload(ctx, fullTombstonePath, bytes.NewReader(data)), "upload tombstone file")
+}
+
+func TombstoneExists(ctx context.Context, bkt objstore.BucketReader, userID string, requestID string, state BlockDeleteRequestState) (bool, error) {
+	// Add the state of the request using the filename extension
+	filename := requestID + "." + string(state) + ".json"
+	fullTombstonePath := path.Join(TombstonePath, filename)
+
+	exists, err := bkt.Exists(ctx, fullTombstonePath)
+
+	if exists || err != nil {
+		return exists, err
+	}
+
+	return false, nil
+}
+
+func GetDeleteRequestByIdForUser(ctx context.Context, bkt objstore.Bucket, cfgProvider bucket.TenantConfigProvider, userID string, requestId string) (*Tombstone, error) {
+	userLogger := util_log.WithUserID(userID, util_log.Logger)
+	userBucket := bucket.NewUserBucketClient(userID, bkt, cfgProvider)
+
+	found := []*Tombstone{}
+
+	for _, state := range AllDeletionStates {
+		filename := requestId + "." + string(state) + ".json"
+
+		exists, err := TombstoneExists(ctx, userBucket, userID, requestId, state)
+		if err != nil {
+			return nil, err
+		}
+
+		if exists {
+			t, err := readTombstoneFile(ctx, userBucket, userID, path.Join(TombstonePath, filename))
+			if err != nil {
+				return nil, err
+			}
+			found = append(found, t)
+		}
+	}
+
+	if len(found) == 0 {
+		return nil, nil
+	}
+
+	// If there are multiple tombstones with the same request id but different state, should delete the files with the lower state
+	for _, ts := range found[0 : len(found)-1] {
+		level.Info(userLogger).Log("msg", "Found extra tombstone file with outdated state. Will delete it.", "requestID", ts.RequestID, "state", ts.State)
+		if err := DeleteTombstoneFile(ctx, bkt, cfgProvider, ts); err != nil {
```

Comment:
I don't like the fact that we are making changes to the bucket during a GET operation. Can't we just perform clean up in the compactor after deletion is complete?

Reply:
Yeah I think that makes more sense. I removed the deletion operations in the get functions.